File: CF Round 953/F_Large_Graph.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <map>
#include <stdint.h>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
using ll = long long;

// https://usaco.guide/gold/dsu
class DisjointSets {

  public:
    vector<int> parents;
    vector<int> sizes;

    DisjointSets(int size) : parents(size), sizes(size, 1) {
        for (int i = 0; i < size; i++) {
            parents[i] = i;
        }
    }

    /** @return the "representative" node in x's component */
    int find(int x) {
        return parents[x] == x ? x : (parents[x] = find(parents[x]));
    }

    /** @return whether the merge changed connectivity */
    bool unite(int x, int y) {
        int x_root = find(x);

        int y_root = find(y);

        if (x_root == y_root) {
            return false;
        }

        if (sizes[x_root] < sizes[y_root]) {
            swap(x_root, y_root);
        }

        sizes[x_root] += sizes[y_root];

        parents[y_root] = x_root;

        return true;
    }

    /** @return whether x and y are in the same connected component */
    bool connected(int x, int y) { return find(x) == find(y); }
};

// https://codeforces.com/blog/entry/62393
static uint64_t x;
uint64_t next() {
    uint64_t z = (x += 0x9e3779b97f4a7c15);
    z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
    z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
    return z ^ (z >> 31);
}
/**
 * Use with
 * unordered_map<long long, int, custom_hash> safe_map;
 * gp_hash_table<long long, int, custom_hash> safe_hash_table;
 * so you don't get hacked :(
 */
struct custom_hash {
    static uint64_t splitmix64(uint64_t x) {
        // http://xorshift.di.unimi.it/splitmix64.c
        x += 0x9e3779b97f4a7c15;
        x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9;
        x = (x ^ (x >> 27)) * 0x94d049bb133111eb;
        return x ^ (x >> 31);
    }

    size_t operator()(uint64_t x) const {
        static const uint64_t FIXED_RANDOM = chrono::steady_clock::now().time_since_epoch().count();
        return splitmix64(x + FIXED_RANDOM);
    }
};

vector<vector<int>> sieve(int n) {
    vector<vector<int>> prime_factors(n + 1, vector<int>(0));
    for (int i = 2; i < n + 1; ++i) {
        if (prime_factors[i].empty()) {      // i.e. if i is prime
            for (int k = i; k < n; k += i) { // everything multiplied by i also has i as a prime factor
                prime_factors[k].push_back(i);
            }
        }
    }
    return prime_factors;
}

auto pf = sieve(1e6 + 5); // precompute the values

void solve() {
    int n, k;
    cin >> n >> k;
    vector<int> a(n);

    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    vector<int> solve(2 * n - 1);

    for (int i = 1; i < n; ++i) {
        solve[i - 1] = a[i];
    }
    for (int i = 0; i < n; ++i) {
        solve[i + n - 1] = a[i];
    }

    DisjointSets dsu(2 * n - 1);

    unordered_map<int, int, custom_hash> lp; // last primes

    for (int i = 0; i < 2 * n - 1; ++i) {
        for (int f : pf[solve[i]]) { // iterate over the factors of e
            if (lp.contains(f) && abs(lp[f] - i) <= k) {
                dsu.unite(lp[f], i);
            }
            lp[f] = i;
        }
    }

    unordered_set<int, custom_hash> u;
    for (auto e : dsu.parents) {
        u.insert(dsu.find(e));
    }
    ll ans = u.size();

    for (int i = 0; i < n; ++i) {
        if (a[i] == 1) {
            if (i == 0) ans += n - 1;
            else ans += n - 2;
        }
    }

    cout << ans << endl;
}

int main() {
    std::ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        solve();
    }
}

File: CF Round 953/F_Large_Graph.py
import sys
from io import BytesIO, IOBase
import os


# sys.setrecursionlimit(10 ** 6)
# Fast IO Region
BUFSIZE = 8192

class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
def input(): return sys.stdin.readline().rstrip("\r\n")
inf = float('inf')
def arr_in():
    return list(map(int, input().split()))

def tup_in():
    return map(int, input().split())

"""
--- Notes ---

"""


class DisjointSets:
	def __init__(self, size: int) -> None:
		self.parents = [i for i in range(size)]
		self.sizes = [1 for _ in range(size)]

	def find(self, x: int) -> int:
		""":return: the "representative" node in x's component"""
		if self.parents[x] == x:
			return x
		self.parents[x] = self.find(self.parents[x])
		return self.parents[x]

	def unite(self, x: int, y: int) -> bool:
		""":return: whether the merge changed connectivity"""
		x_root = self.find(x)
		y_root = self.find(y)
		if x_root == y_root:
			return False

		if self.sizes[x_root] < self.sizes[y_root]:
			x_root, y_root = y_root, x_root

		self.parents[y_root] = x_root
		self.sizes[x_root] += self.sizes[y_root]
		return True

	def connected(self, x: int, y: int) -> bool:
		""":return: whether x and y are in the same connected component"""
		return self.find(x) == self.find(y)


def sieve(n):
    prime_factors = [[] for _ in range(n + 1)]
    for i in range(2, n + 1):
        if not prime_factors[i]: # i.e. if i is prime
            k = 1
            while (i * k < n + 1):
                prime_factors[i * k].append(i)
                k += 1
    return prime_factors

pf = sieve(int(1e6) + 5)

def solution():
    n,k = tup_in()
    a = arr_in()

    s = [0] * (2 * n - 1)

    for i in range(n-1):
        s[i] = a[i+1]
    for i in range(n):
        s[i + n - 1] = a[i]

    dsu = DisjointSets(2 * n - 1)

    lp = dict()

    for i in range(2 * n - 1):
        for f in pf[s[i]]:
            if f in lp and (i-lp[f]) <= k: 
                dsu.unite(lp[f], i)
            lp[f] = i

    ans = len(set(dsu.parents))
    for i in range(n):
         if a[i] == 1:
            if i == 0:
                ans += n - 1
            else:
                ans += n - 2

    print(ans)

if __name__ == "__main__":
    for _ in range(int(input())):
        solution()